src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,421 @@
+    mPassword = NULL;
+    mOAuth2Token = NULL;
+    
+    mImapServices = new Array();
+    mSmtpServices = new Array();
+    mPopServices = new Array();
+    
+    mIdentifier = NULL;
+    mImapServer = NULL;
+    mPopServer = NULL;
+    mSmtpServer = NULL;
+    mImapError = ErrorNone;
+    mPopError = ErrorNone;
+    mSmtpError = ErrorNone;
+    
+    mCurrentServiceIndex = 0;
+    mCurrentServiceTested = 0;
+    
+    mProvider = NULL;
+    
+    mOperation = NULL;
+    mQueue = NULL;
+    mResolveMX = NULL;
+    
+    mImapSession = NULL;
+    mPopSession = NULL;
+    mSmtpSession = NULL;
+}
+
+AccountValidator::AccountValidator()
+{
+    init();
+}
+
+AccountValidator::~AccountValidator()
+{
+    MC_SAFE_RELEASE(mEmail);
+    MC_SAFE_RELEASE(mUsername);
+    MC_SAFE_RELEASE(mPassword);
+    MC_SAFE_RELEASE(mOAuth2Token);
+    MC_SAFE_RELEASE(mImapServices);
+    MC_SAFE_RELEASE(mSmtpServices);
+    MC_SAFE_RELEASE(mPopServices);
+    MC_SAFE_RELEASE(mIdentifier);
+    MC_SAFE_RELEASE(mProvider);
+    MC_SAFE_RELEASE(mOperation);
+    MC_SAFE_RELEASE(mQueue);
+    MC_SAFE_RELEASE(mResolveMX);
+    MC_SAFE_RELEASE(mImapSession);
+    MC_SAFE_RELEASE(mPopSession);
+    MC_SAFE_RELEASE(mSmtpSession);
+}
+
+void AccountValidator::start()
+{
+    if (mEmail == NULL) {
+        if (mUsername == NULL) {
+            return;
+        }
+        else {
+            mEmail = mUsername;
+        }
+    }
+    else if (mUsername == NULL){
+        mUsername = mEmail;
+    }
+    
+    mProvider = MailProvidersManager::sharedManager()->providerForEmail(mUsername);
+    
+    if (mProvider == NULL) {
+        resolveMX();
+    }
+    else{
+        startCheckingHosts();
+    }
+}
+
+void AccountValidator::cancel()
+{
+    if(mOperation != NULL)
+        mOperation->cancel();
+    
+    if(mResolveMX != NULL)
+        mResolveMX->cancel();
+    
+    if (mQueue != NULL)
+        mQueue->cancelAllOperations();
+    
+    MC_SAFE_RELEASE(mImapSession);
+    MC_SAFE_RELEASE(mPopSession);
+    MC_SAFE_RELEASE(mSmtpSession);
+        
+    Operation::cancel();
+}
+
+void AccountValidator::operationFinished(Operation * op)
+{
+    if (op == mResolveMX) {
+        resolveMXDone();
+    }
+    else {
+        checkNextHostDone();
+    }
+}
+
+void AccountValidator::resolveMX()
+{
+    Array * components;
+    String * domain;
+    
+    components = mUsername->componentsSeparatedByString(MCSTR("@"));
+    if (components->count() >= 2) {
+        domain = (String *) components->lastObject();
+        mResolveMX = new MXRecordResolverOperation();
+        mResolveMX->setHostname(domain);
+        mResolveMX->setCallback((OperationCallback *)this);
+        
+        mQueue = new OperationQueue();
+        mQueue->addOperation(mResolveMX);
+    }
+}
+
+void AccountValidator::resolveMXDone()
+{
+    Array * mxRecords = mResolveMX->mxRecords();
+    
+    CFURLRef imageURL = CFBundleCopyResourceURL(CFBundleGetMainBundle(), CFSTR("providers"), CFSTR("json"), NULL);
```

Comment:
Could you just suppose that MailProvidersManager::sharedManager() is already pre-filled? and remove all these lines 162-167.